Added optical elements: custom spectral phase and spectral filter

TimoEichner opened the pull request:
Added optical elements that apply a custom spectral phase (as an array, not polynomial) and a spectral filter. The elements take the phase (or intensity transmission), as well as a frequency axis on which the phase (transmission) is defined, as inputs. When applying to a laser, the phase/transmission is interpolated to the omega grid of the laser.

The elements can be used to spectrally shape laser pulses or apply custom phases, using e.g. measured dispersion/reflectivity curves of optics such as dispersive mirrors.

Example useable of the components:

1. Adding custom spectral phase (in this case a polynomial with GDD and TOD, but could be any phase)

```python
import numpy as np
from scipy.constants import c
from lasy.laser import Laser
from lasy.profiles import GaussianProfile
from lasy.optical_elements import SpectralPhase

w0 = 1e-3
tau = 30e-15/np.sqrt(2*np.log(2))
energy = 1
wavelength = 0.8e-6
pol = (1,0)
t_peak = 0

profile = GaussianProfile(wavelength, pol, laser_energy=energy, w0=w0, tau=tau, t_peak=t_peak)

dim='xyt'
lo = (-5e-3, -5e-3, -500e-15)
hi = (5e-3, 5e-3, 500e-15)
npoints = (100, 100, 200)

# create the laser pulse according to the above specified values
laser = Laser(dim=dim, lo=lo, hi=hi, npoints=npoints, profile=profile)

# set some values of gdd and tod that should be added to the pulse
gdd = 1000e-30
tod = 50000e-45

# calculate the center frequency
omega0 = 2*np.pi*c/wavelength

# create a spectral phase polynomial
phase = 0.5*gdd*(laser.omega_1d-omega0)**2 + 1/6 * tod*(laser.omega_1d-omega0)**3

# define the spectral phase optic
phase = SpectralPhase(phase, laser.omega_1d)

# apply to the laser
laser.apply_optics(phase)

```

2. Applying a SpectralFilter (in this a hard-edged high-pass filter)

```python
import numpy as np
from scipy.constants import c
from lasy.laser import Laser
from lasy.profiles import GaussianProfile
from lasy.optical_elements import SpectralFilter

w0 = 1e-3
tau = 30e-15/np.sqrt(2*np.log(2))
energy = 1
wavelength = 0.8e-6
pol = (1,0)
t_peak = 0

profile = GaussianProfile(wavelength, pol, laser_energy=energy, w0=w0, tau=tau, t_peak=t_peak)

dim='xyt'
lo = (-5e-3, -5e-3, -500e-15)
hi = (5e-3, 5e-3, 500e-15)
npoints = (100, 100, 200)

# create the laser pulse according to the above specified values
laser = Laser(dim=dim, lo=lo, hi=hi, npoints=npoints, profile=profile)

# define cutoff frequency
cutoff = 2*np.pi*c/800e-9

# define the filter array
filter = np.ones_like(laser.omega_1d)

# set transmission at frequencies < omega = 0
filter[laser.omega_1d<cutoff] = 0

# create the SpectralFilter object
filter = SpectralFilter(filter, laser.omega_1d)

# apply to the laser
laser.apply_optics(filter)

```

Commit:
Added optical elements for a custom spectral phase and a spectral filter/spectrally dependent transmission

TimoEichner committed:
Merge branch 'LASY-org:development' into developement

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

TimoEichner committed:
Minor formatting/typo fix

MaxThevenet reviewed (commented):
Looks great, thanks for this PR! I just have 2 minor comments:
1. In the description, for both optics, the arrays are described as `ndarray`s. Can these really be 3D, e.g. (x, y, omega) filter? If not, could you specify this point in the documentation?
2. Could you add a quick example for each, just a few lines of code, maybe as part of the PR description?

TimoEichner committed:
Merge branch 'LASY-org:development' into dev

TimoEichner committed:
Specified required shape of inputs to spectral_filter.py

TimoEichner committed:
Specified required shape of inputs to spectral_phase.py

TimoEichner commented:
Thanks for the comments Maxence!

You're right, the input arrays should be 1D. I added a comment on this to the documentation and added some examples above.

MaxThevenet reviewed (commented):
Perfect! One last thing: could you also add the 2 corresponding files to `docs/source/api/optical_elements`? You could copy what is done e.g. for `Axiparabola`. This will automatically create an entry in the readthedocs documentation [there](https://lasydoc.readthedocs.io/en/latest/api/optical_elements/index.html) for your optical elements.

TimoEichner committed:
Merge branch 'LASY-org:development' into dev

TimoEichner committed:
Create spectral_phase.rst

Added rst file for SpectralPhase to docs

TimoEichner committed:
Create spectral_filter.rst

Added rst file for SpectralFilter to docs

TimoEichner committed:
Update index.rst
